homeassistant/components/aten_pe/switch.py

```diff
@@ -0,0 +1,121 @@
+
+import logging
+import voluptuous as vol
+
+from homeassistant.components.switch import (
+    DEVICE_CLASS_OUTLET,
+    SwitchDevice,
+    PLATFORM_SCHEMA,
+)
+from homeassistant.const import CONF_HOST, CONF_PORT, CONF_USERNAME
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_COMMUNITY = "community"
+CONF_AUTHKEY = "authkey"
+CONF_PRIVKEY = "privkey"
+
+DEFAULT_HOST = "localhost"
+DEFAULT_PORT = "161"
+DEFAULT_COMMUNITY = "private"
+DEFAULT_USERNAME = "administrator"
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Optional(CONF_HOST, default=DEFAULT_HOST): cv.string,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+        vol.Optional(CONF_COMMUNITY, default=DEFAULT_COMMUNITY): cv.string,
+        vol.Optional(CONF_USERNAME, default=DEFAULT_USERNAME): cv.string,
+        vol.Required(CONF_AUTHKEY): cv.string,
+        vol.Required(CONF_PRIVKEY): cv.string,
+    }
+)
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the ATEN PE switch."""
+    from atenpdu import AtenPE
+
+    host = config.get(CONF_HOST)
+    port = config.get(CONF_PORT)
+    community = config.get(CONF_COMMUNITY)
+    username = config.get(CONF_USERNAME)
+    authkey = config.get(CONF_AUTHKEY)
+    privkey = config.get(CONF_PRIVKEY)
+
+    dev = AtenPE(
+        node=host,
+        serv=port,
+        community=community,
+        username=username,
+        authkey=authkey,
+        privkey=privkey,
+    )
+
+    switches = []
+    for outlet in dev.outlets:
+        if outlet.name:
+            switch = AtenSwitch(dev, outlet.id, outlet.name)
+            switches.append(switch)
+
+    async_add_entities(switches)
+    return True
+
+
+class AtenSwitch(SwitchDevice):
+    """Represents an ATEN PE switch."""
+
+    def __init__(self, device, outlet, name):
+        """Initialize an ATEN PE switch."""
+        self._device = device
+        self._outlet = outlet
+        self._name = name
+        self._enabled = False
+        self._outlet_power = 0.0
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return "{mac}-{outlet}".format(mac=self._device.deviceMAC, outlet=self._outlet)
+
+    @property
+    def name(self) -> str:
+        """Return the name of the entity."""
+        return self._name
+
+    @property
+    def device_class(self) -> str:
+        """Return the class of this device, from component DEVICE_CLASSES."""
+        return DEVICE_CLASS_OUTLET
+
+    @property
+    def is_on(self) -> bool:
+        """Return True if entity is on."""
+        return self._enabled
+
+    @property
+    def current_power_w(self) -> float:
+        """Return the current power usage in W."""
+        return self._outlet_power
+
+    def turn_on(self, **kwargs):
+        """Turn the switch on."""
+        self._device.setOutletStatus(self._outlet, "on")
+        self._enabled = True
```

Comment:
Is this needed? The status of the switch is updated right after the change is made.
In general, it should be able to work without this.

Reply:
The PDU lets you configure delays for on and off. It has a third state "pending". But I'd like to avoid this delay in the UI, because it made the visible switch toggle thrice instead of once.